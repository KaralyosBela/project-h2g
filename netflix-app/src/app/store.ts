import { configureStore } from "@reduxjs/toolkit";
import moviesReducer from "../features/moviesSlice";
import hehe from "../features/actions"

export const store = configureStore({
  reducer: {
    movies: moviesReducer,
    bruh: hehe
  }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
