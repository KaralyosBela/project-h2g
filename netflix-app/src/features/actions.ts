export const customAction = (data: string) => {
  return {
    type: "CUSTOM_ACTION",
    payload: data,
  };
};

// Use the initialState as a default value
export default function (state: any = null, action: any) {
  // The reducer normally looks at the action type field to decide what happens
  switch (action.type) {
    case "CUSTOM_ACTION":
      console.log("Custom action finished.");
       return null;
      break;
    default:
      return null;
  }
}
