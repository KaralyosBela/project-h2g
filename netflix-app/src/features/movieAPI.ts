import { createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { IMovies } from "../interfaces/movies.interface";

const moviesAPI = createAsyncThunk(
  "movies/getAllMovies",
  async (args, { rejectWithValue }) => {
    try {
      const response = await axios.get(
        "https://my-json-server.typicode.com/karalyosbela/json-server/movies"
      );

      const data = response.data.map((item: IMovies) => {
        return {
          title: item.title,
          release_date: item.release_date,
          genre: item.genre,
          thumbnail: item.thumbnail,
          movie_url: item.movie_url,
          rating: item.rating,
          runtime: item.runtime,
        };
      });

      return data;
    } catch (error) {
      return rejectWithValue(error);
    }
  }
);
