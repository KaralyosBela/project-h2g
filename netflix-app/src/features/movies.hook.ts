import { useSelector } from "react-redux";
import { IMovies } from "../interfaces/movies.interface";

export const moviesSelector = (state: any) => {
  return state.movies.movies
    .filter((movie: IMovies) =>
      movie.title
        .toLowerCase()
        .includes(state.movies.searchedMovie.toLowerCase())
    )
    .filter(
      (movie: IMovies) =>
        state.movies.filterOptions.genre === "" ||
        movie.genre
          .map((genre) => genre.toLowerCase())
          .includes(state.movies.filterOptions.genre)
    )
    //IMovies nem okÃ©s itt
    .sort((a: any, b: any) => {
      if (state.movies.sortOptions.sortOrder === "ascending") {
        switch (state.movies.sortOptions.sortKey) {
          case "releaseDate":
            return a.release_date - b.release_date;
          case "length":
            return a.runtime.split(" ")[0] - b.runtime.split(" ")[0];
          case "rating":
            return a.rating - b.rating;
        }
      } else {
        switch (state.movies.sortOptions.sortKey) {
          case "releaseDate":
            return b.release_date - a.release_date;
          case "length":
            return b.runtime.split(" ")[0] - a.runtime.split(" ")[0];
          case "rating":
            return b.rating - a.rating;
        }
      }
      return a.release_date - b.release_date;
    });
};

export const useMovies = () => useSelector(moviesSelector);
