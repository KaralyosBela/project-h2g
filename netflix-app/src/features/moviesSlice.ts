import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";
import { IMovie } from "../interfaces/movies.interface";

export const getMovies = createAsyncThunk(
  "movies/getMovies",
  async (args, { rejectWithValue }) => {
    try {
      const response = await axios.get("http://localhost:4000/movies");

      const movieData: IMovie = response.data.data.map((item: IMovie) => {
        return {
          id: item.id,
          title: item.title,
          tagline: item.tagline,
          vote_average: item.vote_average,
          budget: item.budget,
          revenue: item.revenue,
          release_date: item.release_date.slice(0, 4),
          genres: item.genres,
          poster_path: item.poster_path,
          vote_count: item.vote_count,
          runtime: item.runtime,
          overview: item.overview,
        };
      });
      console.log(movieData);
      return movieData;
    } catch (error: unknown) {
      return rejectWithValue("Fetch failed!");
    }
  }
);

// export const addMovie = async (movie: IMovieWithoutID) => (dispatch: any) => {
//   try {
//     //post without ID
//     axios.post("http://localhost:4000/movies", {
//       title: movie.title,
//       tagline: movie.tagline,
//       vote_average: movie.vote_average,
//       budget: movie.budget,
//       revenue: movie.revenue,
//       release_date: movie.release_date,
//       genres: movie.genres,
//       poster_path: movie.poster_path,
//       vote_count: movie.vote_count,
//       runtime: movie.runtime,
//       overview: movie.overview,
//     });
//     dispatch(getMovies());
//     // return movie;
//   } catch (error) {}
// };

export const addMovie = createAsyncThunk(
  "movies/addMovie",
  async (movie: IMovie, { rejectWithValue }) => {
    console.log(movie);
    try {
      axios.post("http://localhost:4000/movies", {
        title: movie.title,
        tagline: movie.tagline,
        vote_average: movie.vote_average,
        budget: movie.budget,
        revenue: movie.revenue,
        release_date: movie.release_date,
        genres: movie.genres,
        poster_path: movie.poster_path,
        vote_count: movie.vote_count,
        runtime: movie.runtime,
        overview: movie.overview,
      });
      return movie;
    } catch (error: unknown) {
      return rejectWithValue("Post failed!");
    }
  }
);

export const deleteMovie = createAsyncThunk(
  "movies/deleteMovie",
  async (movie: IMovie, { rejectWithValue }) => {
    try {
      axios.delete(`http://localhost:4000/movies/${movie.id}`);
      return movie;
    } catch (error: unknown) {
      return rejectWithValue(error);
    }
  }
);

export const editMovie = createAsyncThunk(
  "movies/editMovie",
  async (movie: IMovie, { rejectWithValue }) => {
    console.log(movie);
    try {
      axios.put(`http://localhost:4000/movies`, movie);
      return movie;
    } catch (error: unknown) {
      return rejectWithValue("Edit failed!");
    }
  }
);

interface moviesState {
  movies: IMovie[];
  searchedMovie: string;
  movie: IMovie;
  bannerVisible: boolean;
  sortOptions: {
    sortKey: string;
    sortOrder: string;
  };
  filterOptions: {
    genre: string;
  };
  numberOfActionMovies: number;
}

const initialState: moviesState = {
  movies: [],
  searchedMovie: "",
  bannerVisible: false,
  movie: {
    tagline: "",
    vote_average: 0,
    vote_count: 0,
    poster_path: "",
    budget: 0,
    revenue: 0,
    id: "",
    genres: [],
    release_date: "",
    runtime: 0,
    title: "",
    overview: "",
  },
  sortOptions: {
    sortKey: "",
    sortOrder: "",
  },
  filterOptions: {
    genre: "",
  },
  numberOfActionMovies: 0,
};

export const moviesSlice = createSlice({
  name: "movies",
  initialState,
  reducers: {
    setSearchedMovie: (state, action: PayloadAction<string>) => {
      state.searchedMovie = action.payload;
    },
    setGenreFilter: (state, action: PayloadAction<string>) => {
      state.filterOptions.genre = action.payload;
    },
    setSortParams: (
      state,
      action: PayloadAction<{ key: string; order: string }>
    ) => {
      state.sortOptions.sortKey = action.payload.key;
      state.sortOptions.sortOrder = action.payload.order;
    },
    setChoosenMovie: (state, action: PayloadAction<IMovie>) => {
      state.movie = action.payload;
    },
    setMovieBannerStatus: (state, action: PayloadAction<boolean>) => {
      state.bannerVisible = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMovies.fulfilled, (state, action: PayloadAction<IMovie>) => {
        state.movies = Object.values(action.payload);
      })
      .addCase(getMovies.rejected, (state, action: any) => {
        console.log(action.payload);
      })
      .addCase(addMovie.fulfilled, (state, action: PayloadAction<IMovie>) => {
        state.movies.push(action.payload);
      })
      .addCase(addMovie.rejected, (state, action: any) => {
        console.log(action.payload);
      })
      .addCase(
        deleteMovie.fulfilled,
        (state, action: PayloadAction<IMovie>) => {
          state.movies = state.movies.filter(
            (movie) => movie.id !== action.payload.id
          );
        }
      )
      .addCase(deleteMovie.rejected, (state, action: any) => {
        console.log(action.payload);
      })
      .addCase(editMovie.fulfilled, (state, action: PayloadAction<IMovie>) => {
        const index = state.movies.findIndex(
          (movie) => movie.id === action.payload.id
        );
        state.movies[index] = action.payload;
      })
      .addCase(editMovie.rejected, (state, action: any) => {
        console.log(action.payload);
      });
  },
});

export const {
  setSearchedMovie,
  setGenreFilter,
  setSortParams,
  setChoosenMovie,
  setMovieBannerStatus,
} = moviesSlice.actions;
export default moviesSlice.reducer;
