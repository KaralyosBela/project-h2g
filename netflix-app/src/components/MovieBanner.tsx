import { IMovie } from "../interfaces/movies.interface";
import classes from "./MovieBanner.module.css";
import {FaSearch } from "react-icons/fa"
import { setMovieBannerStatus } from "../features/moviesSlice";
import { AppDispatch } from "../app/store";
import { useDispatch } from "react-redux";

interface Props {
  movie: IMovie;
}

export const MovieBanner: React.FC<Props> = ({ movie }) => {
  const dispatch = useDispatch<AppDispatch>();

  const setMovieBannerState = () => {
    dispatch(setMovieBannerStatus(false));
  }

  return (
    <div className={classes.movieBanner}>
      <div className={classes.titleBar}>
        <h3 className={classes.pageTitle}>netflixRoulette</h3>
        <div className={classes.searchIcon} onClick={setMovieBannerState}><FaSearch size={30}/></div>
      </div>
      <div className={classes.infoPart}>
        <img className={classes.image} src={movie.poster_path} alt="alt"></img>
        <div className={classes.movieInfo}>
          <div className={classes.firstRow}>
            <h1 className={classes.title}>{movie.title}</h1>
            <h1 className={classes.vote_average}>
            {movie.vote_average}
            </h1>
          </div>
          <p className={classes.genre}>{movie.genres.join(", ")}</p>
          <div className={classes.secondRow}>
            <h3 className={classes.release_date}>{movie.release_date}</h3>
            <h3 className={classes.runtime}>{movie.runtime} min</h3>
          </div>
          <div className={classes.thirdRow}>
            <p className={classes.overview}>{movie.overview}</p>
          </div>
        </div>
      </div>
    </div>
  );
};
