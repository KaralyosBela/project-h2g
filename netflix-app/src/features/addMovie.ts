import axios from "axios";
import { IMovieWithoutID } from "../interfaces/movies.interface";
import { getMovies } from "./moviesSlice";

export const addMovie = (movie: IMovieWithoutID, dispatch: any) => {
    try {
        axios.post("http://localhost:4000/movies", {
          title: movie.title,
          tagline: movie.tagline,
          vote_average: movie.vote_average,
          budget: movie.budget,
          revenue: movie.revenue,
          release_date: movie.release_date,
          genres: movie.genres,
          poster_path: movie.poster_path,
          vote_count: movie.vote_count,
          runtime: movie.runtime,
          overview: movie.overview,
        });
        console.log("object");
        dispatch(getMovies());
      } catch (error) {}
    }