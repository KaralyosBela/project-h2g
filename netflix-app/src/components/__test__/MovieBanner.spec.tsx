import { fireEvent, render, screen } from "@testing-library/react";
import { store } from "../../app/store";
import { setMovieBannerStatus } from "../../features/moviesSlice";
import { MovieBanner } from "../MovieBanner";

const movie = {
  id: "903",
  title: "Cool Hand Luke",
  tagline: "What we've got here is failure to communicate.",
  vote_average: 7.7,
  vote_count: 471,
  release_date: "1967-06-22",
  poster_path:
    "https://image.tmdb.org/t/p/w500/vDwphkloD7ToaDpKASAXGgHOclN.jpg",
  overview:
    "When petty criminal Luke Jackson is sentenced to two years in a Florida prison farm, he doesn't play by the rules of either the sadistic warden or the yard's resident heavy, Dragline, who ends up admiring the new guy's unbreakable will. Luke's bravado, even in the face of repeated stints in the prison's dreaded solitary confinement cell, \"the box,\" make him a rebel hero to his fellow convicts and a thorn in the side of the prison officers.",
  budget: 3200000,
  revenue: 16217773,
  genres: ["Crime", "Drama"],
  runtime: 126,
};

const mockDispatch = jest.fn();
const mockUseDispatch = jest.fn();
jest.mock("react-redux", () => ({ useDispatch: () => mockDispatch }));
mockDispatch.mockReturnValue(mockUseDispatch);

describe("Movie banner component", () => {
  it("should render movie banner", () => {
    const { container } = render(<MovieBanner movie={movie} />);
    expect(container).toMatchSnapshot();
  });

  it("should dispatch on 'searchIcon' click", () => {
    render(<MovieBanner movie={movie} />);
    const searchIcon = screen.getByTestId("searchIcon");
    fireEvent.click(searchIcon);
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(setMovieBannerStatus(false));
  });
});
