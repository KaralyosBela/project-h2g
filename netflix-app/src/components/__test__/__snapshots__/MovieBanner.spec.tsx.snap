// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie banner component should render movie banner 1`] = `
<div>
  <div
    class="movieBanner"
  >
    <div
      class="titleBar"
    >
      <h3
        class="pageTitle"
      >
        netflixRoulette
      </h3>
      <div
        class="searchIcon"
      >
        <svg
          fill="currentColor"
          height="30"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </div>
    </div>
    <div
      class="infoPart"
    >
      <img
        alt="alt"
        class="image"
        src="https://image.tmdb.org/t/p/w500/vDwphkloD7ToaDpKASAXGgHOclN.jpg"
      />
      <div
        class="movieInfo"
      >
        <div
          class="firstRow"
        >
          <h1
            class="title"
          >
            Cool Hand Luke
          </h1>
          <span
            class="vote_average"
          >
            7.7
          </span>
        </div>
        <p
          class="genre"
        >
          Crime, Drama
        </p>
        <div
          class="secondRow"
        >
          <h3
            class="release_date"
          >
            1967-06-22
          </h3>
          <h3
            class="runtime"
          >
            126
             min
          </h3>
        </div>
        <div
          class="thirdRow"
        >
          <p
            class="overview"
          >
            When petty criminal Luke Jackson is sentenced to two years in a Florida prison farm, he doesn't play by the rules of either the sadistic warden or the yard's resident heavy, Dragline, who ends up admiring the new guy's unbreakable will. Luke's bravado, even in the face of repeated stints in the prison's dreaded solitary confinement cell, "the box," make him a rebel hero to his fellow convicts and a thorn in the side of the prison officers.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
